[SMA] Improving the Parallel rule of mixtures

AlejandroCornejo opened the pull request:
In this PR I'm improving the existing rule of mixtures allowing to orient each layer according to a set of Euler angles and enhancing the tangent tensor calculation for non-linear components.

AlejandroCornejo committed:
improving

loumalouomega committed:
Merge branch 'master' into structural/improving-parallel-rule-of-mixtures

loumalouomega committed:
Update authorship

loumalouomega committed:
Inconsistent deformation gradient

AlejandroCornejo committed:
missing flag

AlejandroCornejo committed:
g

loumalouomega committed:
Clean up

AlejandroCornejo committed:
adding variable

AlejandroCornejo committed:
adding methods

AlejandroCornejo committed:
more

AlejandroCornejo committed:
1st implementation

AlejandroCornejo committed:
adding

AlejandroCornejo committed:
now working

AlejandroCornejo committed:
Merge branch 'structural/adding-implicit-anisotropic-generic-law' into structural/improving-parallel-rule-of-mixtures

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
kirchoff ok

AlejandroCornejo committed:
found tricky error!

AlejandroCornejo committed:
templating dimension and change name

AlejandroCornejo committed:
completing

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
optimizing anisotropic cl

AlejandroCornejo committed:
final changes

AlejandroCornejo committed:
cl improved DONE

AlejandroCornejo commented:
BTW, this PR should be merged after #6836

philbucher commented on applications/StructuralMechanicsApplication/custom_advanced_constitutive/rule_of_mixtures_law.h:
```diff
@@ -51,8 +53,10 @@ namespace Kratos
+ * @author Alejandro Cornejo Velazquez
  */
-class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) RuleOfMixturesLaw
+template<unsigned int TDim>
```
another template? :sweat_smile: :sweat_smile: :sweat_smile:

philbucher replied:
This is TBD (@RiccardoRossi)
the compilation and linking times of the CLaws that are already heavily templated is an issue for a while already

AlejandroCornejo replied:
yeah... In this case I need to know the Dimension of the problem if using BoundedMatrices for the rotations

philbucher commented on applications/StructuralMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -21,6 +21,8 @@
 #include "includes/node.h"
 #include "geometries/geometry.h"
+#include "includes/constitutive_law.h"
+#include "structural_mechanics_application_variables.h"
```
neither of those includes is used in this file, please remove them

AlejandroCornejo replied:
the one with the variables has to be included, to other has been erased, THX

AlejandroCornejo committed:
avoid compilation warning (MACELI)

philbucher commented on applications/StructuralMechanicsApplication/structural_mechanics_application.cpp:
```diff
@@ -405,6 +405,7 @@ void KratosStructuralMechanicsApplication::Register() {
     KRATOS_REGISTER_VARIABLE(PRESCRIBED_DISPLACEMENT)
 
     // Constitutive laws variables
+    KRATOS_REGISTER_VARIABLE(LAYER_EULER_ANGLES)
```
see my comment in the other PR, please refactor the CL-vars into a separate file

AlejandroCornejo replied:
Yes! I saw your comment. The thing is that I agree but I wouldn't do that in one of these PR's. If you want I can create an issue and discuss this. Once we all agree I can implement this refactor. @RiccardoRossi @loumalouomega

philbucher replied:
there is already #3874
but nothing happened in almost 1.5 years

philbucher replied:
in any case we can do it in a different PR, I will release my block once someone else (@loumalouomega) took a closer look

loumalouomega replied:
I agree with this PR, but the anisotropic PR must be merged first

philbucher reviewed (changes requested):
I didn't review in detail yet

AlejandroCornejo committed:
erasing include

loumalouomega commented:
CI is failing

AlejandroCornejo commented:
> CI is failing

I have to update your test I suppose

loumalouomega committed:
Merge branch 'master' into structural/improving-parallel-rule-of-mixtures

AlejandroCornejo committed:
Merge branch 'master' into structural/improving-parallel-rule-of-mixtures

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
update test

philbucher commented:
I see I am blocking
what is the status?
CI is failing

AlejandroCornejo commented:
yes, I have to correct the CL and should be fine

AlejandroCornejo committed:
name test corrected

AlejandroCornejo commented:
@loumalouomega I've changed the name of the `RuleOfMixturesLaw` by `ParallelRuleOfMixtures` but somehow the test are ot detecting this and the CL fails. I've changed the name of the law inside the cpp_test. Do you know what's going on?

loumalouomega commented:
> 
> 
> @loumalouomega I've changed the name of the `RuleOfMixturesLaw` by `ParallelRuleOfMixtures` but somehow the test are ot detecting this and the CL fails. I've changed the name of the law inside the cpp_test. Do you know what's going on?

Looks like is not registered

loumalouomega commented on applications/StructuralMechanicsApplication/structural_mechanics_application.cpp:
```diff
@@ -1122,7 +1123,8 @@ void KratosStructuralMechanicsApplication::Register() {
     // Rules of mixtures
-    KRATOS_REGISTER_CONSTITUTIVE_LAW("RuleOfMixturesLaw", mRuleOfMixturesLaw);
+    KRATOS_REGISTER_CONSTITUTIVE_LAW("ParallelRuleOfMixturesLaw2D", mParallelRuleOfMixturesLaw2D);
+    KRATOS_REGISTER_CONSTITUTIVE_LAW("ParallelRuleOfMixturesLaw3D", mParallelRuleOfMixturesLaw3D);
```
You have changed the name to ParallelRuleOfMixturesLaw3D and ParallelRuleOfMixturesLaw2D, change the Parameters of the test to ParallelRuleOfMixturesLaw3D instead of ParallelRuleOfMixturesLaw

loumalouomega replied:
@AlejandroCornejo

AlejandroCornejo replied:
I've done it!! :S

loumalouomega replied:
![imagen](https://user-images.githubusercontent.com/19991680/83252967-748bfe80-a1ac-11ea-989b-011649fcd191.png)

AlejandroCornejo commented on applications/StructuralMechanicsApplication/tests/cpp_advanced_constitutive_tests/test_rule_of_mixtures_law.cpp:
```diff
@@ -55,7 +55,7 @@ Parameters GetTwoLayersParameters()
         "Material"        : {
             "constitutive_law" : {
-                "name" : "ParallelRuleOfMixturesLaw",
+                "name" : "ParallelRuleOfMixturesLaw3D",
```
I've changed the name here! should be fine

loumalouomega replied:
Check three layers:
![imagen](https://user-images.githubusercontent.com/19991680/83252996-82418400-a1ac-11ea-8ff2-61b6ad4e1eaa.png)

AlejandroCornejo commented on applications/StructuralMechanicsApplication/tests/cpp_advanced_constitutive_tests/test_rule_of_mixtures_law.cpp:
```diff
@@ -104,7 +104,7 @@ Parameters GetThreeLayersParameters()
         "Material"        : {
             "constitutive_law" : {
-                "name" : "ParallelRuleOfMixturesLaw",
+                "name" : "ParallelRuleOfMixturesLaw3D",
```
and here the 3 layers

loumalouomega committed:
Merge branch 'master' into structural/improving-parallel-rule-of-mixtures

loumalouomega commented:
@AlejandroCornejo looks like has been solved after merging the master ... ¯\_(ツ)_/¯

philbucher commented:
tests are still failing

AlejandroCornejo committed:
minor mistake

AlejandroCornejo committed:
update test and correct minor mistake

AlejandroCornejo commented:
Okkey guys, I've updated the test and now it should be compiling. @philbucher this PR includes the correction of the noisy warning in the rule of mixtures :)

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_advanced_constitutive_tests/test_rule_of_mixtures_law.cpp:
```diff
@@ -30,21 +30,21 @@ namespace Testing
-//     gid_io.PrintOnGaussPoints(PK2_STRESS_VECTOR, ThisModelPart, label);
-//     gid_io.WriteNodalFlags(ACTIVE, "ACTIVE", ThisModelPart.Nodes(), label);
-// }
+void GiDIODebugRuleMixtures(ModelPart& ThisModelPart)
```
@AlejandroCornejo please comment this, this should not be enabled in master

AlejandroCornejo committed:
gid_io commented in test

philbucher reviewed (dismissed):
the remaining things I pointed out can/should be discussed separately

thx for fixing the warning

philbucher commented:
codacy has a valid suggestion

AlejandroCornejo committed:
codacy suggestion

AlejandroCornejo committed:
another codacy
